examples/model_compress/pruning/pruning_bert_glue.py

```diff
@@ -0,0 +1,355 @@
+
+    distil_training = functools.partial(training, teacher_model=teacher_model, log_path=log_dir / 'taylor_pruning.log')
+    traced_optimizer = nni.trace(Adam)(attention_pruned_model.parameters(), lr=3e-5, eps=1e-8)
+    evaluator = TorchEvaluator(distil_training, traced_optimizer, fake_criterion)
+
+    attention_pruned_model.train()
+    current_step = 0
+    best_result = 0
+    total_epochs = 4
+    init_lr = 3e-5
+
+    for current_epoch in range(total_epochs):
+        for batch in train_dataloader:
+            # pruning 12 times
+            if current_step % 2000 == 0 and current_step < 24000:
+                check_point = attention_pruned_model.state_dict()
+                pruner = TaylorFOWeightPruner(attention_pruned_model, ffn_config_list, evaluator, 1000)
+                _, ffn_masks = pruner.compress()
+                renamed_ffn_masks = {}
+                for model_name, targets_mask in ffn_masks.items():
+                    renamed_ffn_masks[model_name.split('bert.encoder.')[1]] = targets_mask
+                pruner._unwrap_model()
+                attention_pruned_model.load_state_dict(check_point)
+                ModelSpeedup(attention_pruned_model.bert.encoder, torch.rand(8, 128, 768).to(device), renamed_ffn_masks).speedup_model()
+                optimizer = Adam(attention_pruned_model.parameters(), lr=init_lr)
+            batch.to(device)
+            optimizer.zero_grad()
+
+            # manually schedule lr
+            for params_group in optimizer.param_groups:
+                params_group['lr'] = (1 - current_step / (total_epochs * steps_per_epoch)) * init_lr
+
+            outputs = attention_pruned_model(**batch)
+            loss = outputs.loss
+
+            # distillation
+            if teacher_model:
+                distil_loss = F.kl_div(F.log_softmax(outputs.logits / 2, dim=-1), F.softmax(teacher_model(**batch).logits / 2, dim=-1), reduction='batchmean') * (2 ** 2)
+                loss = 0.1 * loss + 0.9 * distil_loss
+            loss.backward()
+            optimizer.step()
+
+            current_step += 1
+            if current_step % 1000 == 0 or current_step % len(train_dataloader) == 0:
+                result = evaluation(attention_pruned_model)
+                with (log_dir / 'ffn_pruning.log').open('a+') as f:
+                    f.write('[{}] Epoch {}, Step {}: {}\n'.format(time.asctime(time.localtime(time.time())), current_epoch, current_step, result))
+                if current_step >= 24000 and best_result < result[0]:
+                    torch.save(attention_pruned_model, log_dir / 'best_model.pth')
+                    best_result = result[0]
+                    with (log_dir / 'ffn_pruning.log').open('a+') as f:
+                        original_stdout = sys.stdout
+                        sys.stdout = f
+                        print(attention_pruned_model)
+                        sys.stdout = original_stdout
```

Comment:
though this example has good performance, it is not that user friendly, we should think about how to make it user friendly while providing enough flexibility. but not in this pr.

Reply:
good suggestion, I will seriously consider about this